marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mastering/default/merging.sjs

```diff
@@ -39,86 +38,85 @@ function main(content, options) {
-    ).toArray()
-        // get the actionNode for the URI
-        .map(
-          (result) => result.xpath(`/matchSummary/actionDetails/*[action = 'merge']`).toArray()
-            .filter((actionNode) => cts.contains(actionNode, urisQuery))[0]
-        )
-        // filter out false positives that don't have merge actions for the given URI
-        .filter((actionNode) => !!actionNode)
-        // convert node to JSON object
-        .map((actionNode) => actionNode.toObject());
-    if (otherMergesForURI.some((otherMerge) => otherMerge.uris.length > uriActionDetails.uris.length)) {
-      return Sequence.from([]);
+
+  let returnArray = [];
+
+  let thisMatchSummary = (content.value ? content.value : cts.doc(thisMatchSummaryURI));
+  if (!thisMatchSummary) {
+    return Sequence.from([]);
+  }
+  thisMatchSummary = thisMatchSummary.toObject();
+  let theseURIsToProcess = thisMatchSummary.matchSummary.URIsToProcess;
+
+  for (uriToProcess of theseURIsToProcess) {
+    // don't process this "URIsToProcess" URI unless we are processing the last matchSummary document
+    // (in URI order) that contains it
+    let lastSummaryWithURI = cts.uris(null, ["descending", "limit=1"],
+      cts.pathRangeQuery("/matchSummary/URIsToProcess", "=" , uriToProcess));
+    if (lastSummaryWithURI != thisMatchSummaryURI) {
+      continue;
     }
-    otherMergesForURI.forEach((actionDetails) => actionDetails.uris.forEach(
-      (uri) => {
-        if (!uriActionDetails.uris.includes(uri)) {
-          uriActionDetails.uris.push(uri);
-        }
+
+    let uriActionDetails = thisMatchSummary.matchSummary.actionDetails[uriToProcess];
+    // If the action is merge, ensure we create the merge with the most URIs
+    if (uriActionDetails && uriActionDetails.action === 'merge') {
+      const urisQuery = cts.pathRangeQuery('//actionDetails/*/uris', '=', uriActionDetails.uris)
+      const positiveQuery =
+        cts.andQuery([
+          cts.jsonPropertyValueQuery('action', 'merge'),
+          urisQuery,
+          collectionQuery
+        ]);
+      const otherMergesForURI =
+        cts.search(
+          cts.andNotQuery(
+            positiveQuery,
+            cts.rangeQuery(urisPathReference, '=', uriToProcess)
+          ),
+          ['unfiltered',cts.indexOrder(datahubCreatedOnRef, 'descending')]
+        ).toArray()
+          // get the actionNode for the URI
+          .map(
+            (result) => result.xpath(`/matchSummary/actionDetails/*[action = 'merge']`).toArray()
+              .filter((actionNode) => cts.contains(actionNode, urisQuery))[0]
+          )
+          // filter out false positives that don't have merge actions for the given URI
+          .filter((actionNode) => !!actionNode)
+          // convert node to JSON object
+          .map((actionNode) => actionNode.toObject());
+      if (otherMergesForURI.some((otherMerge) => otherMerge.uris.length > uriActionDetails.uris.length)) {
+        return Sequence.from([]);
```

Comment:
I believe this could cause us to miss out on other URIs in the match summary that doesn't have more recent updates in other match summaries. 

It would probably be better to filter `theseURIsToProcess` with this logic and then make a single call to `buildContentObjectsFromMatchSummary` with the sequence of URIs in that filtered list.